File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/equal_partition.py
from __future__ import annotations

from dp_algorithms import subset_sum
from dp_algorithms.common import DPResult


def memoized(nums: list[int]) -> DPResult:
    total = sum(nums)
    value = total % 2 == 0 and bool(subset_sum.memoized(nums, total // 2).value)
    return DPResult("Equal Partition", "memoization", value)


def tabulated(nums: list[int]) -> DPResult:
    total = sum(nums)
    if total % 2 != 0:
        return DPResult("Equal Partition", "tabulation", False, {"target": None, "subset": []})
    result = subset_sum.tabulated(nums, total // 2)
    return DPResult("Equal Partition", "tabulation", result.value, {"target": total // 2, "subset": result.details["subset"]})


def space_optimized(nums: list[int]) -> DPResult:
    total = sum(nums)
    value = total % 2 == 0 and bool(subset_sum.space_optimized(nums, total // 2).value)
    return DPResult("Equal Partition", "space optimized", value)

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/EqualPartition.java
package com.example.dp;

import java.util.Map;

public final class EqualPartition {
    private EqualPartition() {
    }

    public static Result memoized(int[] nums) {
        int total = sum(nums);
        boolean value = total % 2 == 0 && Boolean.TRUE.equals(SubsetSum.memoized(nums, total / 2).value());
        return new Result("Equal Partition", "memoization", value, Map.of());
    }

    public static Result tabulated(int[] nums) {
        int total = sum(nums);
        if (total % 2 != 0) {
            return new Result("Equal Partition", "tabulation", false, Map.of("target", "none"));
        }
        Result result = SubsetSum.tabulated(nums, total / 2);
        return new Result("Equal Partition", "tabulation", result.value(), Map.of("target", total / 2, "subset", result.details().get("subset")));
    }

    public static Result spaceOptimized(int[] nums) {
        int total = sum(nums);
        boolean value = total % 2 == 0 && Boolean.TRUE.equals(SubsetSum.spaceOptimized(nums, total / 2).value());
        return new Result("Equal Partition", "space optimized", value, Map.of());
    }

    private static int sum(int[] nums) {
        int total = 0;
        for (int num : nums) {
            total += num;
        }
        return total;
    }
}
